core/server/web/shared/middlewares/validation/upload.js

```diff
@@ -17,18 +17,26 @@ module.exports = function upload(options) {
 
         // Check if a file was provided
         if (!localUtils.checkFileExists(req.file)) {
-            return next(new common.errors.ValidationError({
+            const err = new common.errors.ValidationError({
                 message: common.i18n.t(`errors.api.${type}.missingFile`)
-            }));
+            });
+            const resource = type.replace(/s$/, '');
+            err.userMessage = common.i18n.t('errors.api.userMessages.ValidationError', {action: `upload ${resource}`});
```

Comment:
This is a very hacky tbh. Had to put it here because we can't control validation errors that happen outside of the frame pipeline in places like https://github.com/TryGhost/Ghost/blob/master/core/server/web/api/v2/admin/routes.js#L198-L199. The best solution for this would be if we somehow could move upload validation inside the frame pipeline, but for now there's no better place without bigger changes.

Reply:
@kirrg001 fyi ^

Reply:
Similar thing goes for authorization. Now questioning the retional of putting this `userMessage` handling inside of frame..

Reply:
hm yeah it feels like it misses a design.

mm can't we just handle the user error message completely in the error handler?

based on
- http method
- url (contains the resource name)

?

Reply:
If we move the logic into the error handler it will tie too much to the routing and extracting just the resource name could be not enough information (e.g. we might need action information about the action like for `authentication` resource there are couple actions that map to POST: `acceptInvitation` and `setup`)

Reply:
Yeah there must be a differentiation between resource endpoints and others.

I agree the validation middleware probably needs to live in the API validation layer and not outside of it.

This needs merging asap and I would raise an issue to re-think how we generate the user error messages in a nicer way.